### PR TITLE
Fix builder inconsistent error on buggy platform

### PR DESCRIPTION
**Pull request by cpuguy83**
addresses https://github.com/docker/for-linux/issues/1170

When pulling an image by platform, it is possible for the image's
configured platform to not match what was in the manifest list.
The image itself is buggy because either the manifest list is incorrect
or the image config is incorrect. In any case, this is preventing people
from upgrading because many times users do not have control over these
buggy images.

This was not a problem in 19.03 because we did not compare on platform
before. It just assumed if we had the image it was the one we wanted
regardless of platform, which has its own problems.

Example Dockerfile that has this problem:

```Dockerfile
FROM --platform=linux/arm64 k8s.gcr.io/build-image/debian-iptables:buster-v1.3.0
RUN echo hello
```

This fails the first time you try to build after it finishes pulling but
before performing the `RUN` command.
On the second attempt it works because the image is already there and
does not hit the code that errors out on platform mismatch (Actually it
ignores errors if an image is returned at all).

Must be run with the classic builder (DOCKER_BUILDKIT=0).

This should mostly be safe (and certainly safer than 19.03) since we only check this immediately after a pull.

Example output before the fix:

```
root@07d110f25ff1:/go/src/github.com/docker/docker# DOCKER_BUILDKIT=0 docker build -t test -< ./Dockerfile.fix 
Sending build context to Docker daemon  2.048kB
Step 1/2 : FROM --platform=linux/arm64 k8s.gcr.io/build-image/debian-iptables:buster-v1.3.0
buster-v1.3.0: Pulling from build-image/debian-iptables
0b711604db4a: Pull complete 
952e2ac66b7a: Pull complete 
37e66e9a979e: Pull complete 
c3cbd1f2e855: Pull complete 
0e1a7dafc259: Pull complete 
643e4c7dc28d: Pull complete 
Digest: sha256:4c9410a4ee555dcb0e8b7bd6fc77c65ac400f7c5bd4555df8187630efaea6ea4
Status: Downloaded newer image for k8s.gcr.io/build-image/debian-iptables:buster-v1.3.0
image with reference k8s.gcr.io/build-image/debian-iptables:buster-v1.3.0 was found but does not match the specified platform: wanted linux/arm64, actual: linux/amd64
root@07d110f25ff1:/go/src/github.com/docker/docker# 
```

Example output after the fix:

```
root@34d494e751a0:/go/src/github.com/docker/docker# DOCKER_BUILDKIT=0 docker build -t test -< ./Dockerfile.fix 
Sending build context to Docker daemon  2.048kB
Step 1/2 : FROM --platform=linux/arm64 k8s.gcr.io/build-image/debian-iptables:buster-v1.3.0
buster-v1.3.0: Pulling from build-image/debian-iptables
0b711604db4a: Pull complete 
952e2ac66b7a: Pull complete 
37e66e9a979e: Pull complete 
c3cbd1f2e855: Pull complete 
0e1a7dafc259: Pull complete 
643e4c7dc28d: Pull complete 
Digest: sha256:4c9410a4ee555dcb0e8b7bd6fc77c65ac400f7c5bd4555df8187630efaea6ea4
Status: Downloaded newer image for k8s.gcr.io/build-image/debian-iptables:buster-v1.3.0

WARNING: Pulled image with specified platform (linux/arm64), but resulting image's configured platform (linux/amd64) does not match.
This is most likely caused by a bug in the build system that created the fetched image (k8s.gcr.io/build-image/debian-iptables:buster-v1.3.0).
Please notify the image author to correct the configuration.

 ---> ee124318448e
Step 2/2 : RUN echo hello
 ---> Running in 375eb8439033
hello
Removing intermediate container 375eb8439033
 ---> cd1060cadbce
Successfully built cd1060cadbce
Successfully tagged test:latest
```

### CODE REVIEW

File: daemon/images/image_builder.go [hunk 1]
```diff
@@ -161,7 +167,27 @@ func (i *ImageService) pullForBuilder(ctx context.Context, name string, authConf
...
+
+		p := *platform
+		if !platforms.Only(p).Match(imgPlat) {
+			po := streamformatter.NewJSONProgressOutput(output, false)
+			progress.Messagef(po, "", `
+WARNING: Pulled image with specified platform (%s), but the resulting image's configured platform (%s) does not match.
+This is most likely caused by a bug in the build system that created the fetched image (%s).
+Please notify the image author to correct the configuration.
+				`, platforms.Format(p), platforms.Format(imgPlat), name)
+			logrus.WithError(err).WithField("image", name).Warn("Ignoring error about platform mismatch where the manifest list points to an image whose configuration does not match the platform in the manifest.")
```
Comment (thaJeztah) [inline daemon/images/image_builder.go:186]: Think it's worth logging image platform as well?

Comment (cpuguy83) [reply daemon/images/image_builder.go:186]: We can, but it is in the error message itself which we already have.

Comment (thaJeztah) [reply daemon/images/image_builder.go:186]: ah, yes, makes sense

**Review by thaJeztah — APPROVED** [review]
SGTM

@tonistiigi @tianon PTAL

File: daemon/images/image_builder.go [hunk 2]
```diff
@@ -161,7 +167,27 @@ func (i *ImageService) pullForBuilder(ctx context.Context, name string, authConf
...
+		}
+
+		p := *platform
+		if !platforms.Only(p).Match(imgPlat) {
+			po := streamformatter.NewJSONProgressOutput(output, false)
+			progress.Messagef(po, "", `
+WARNING: Pulled image with specified platform (%s), but the resulting image's configured platform (%s) does not match.
+This is most likely caused by a bug in the build system that created the fetched image (%s).
+Please notify the image author to correct the configuration.
+				`, platforms.Format(p), platforms.Format(imgPlat), name)
```
Comment (thaJeztah) [inline daemon/images/image_builder.go:185]:
```suggestion
`, platforms.Format(p), platforms.Format(imgPlat), name)
```

**Review by thaJeztah — APPROVED** [review]
still LGTM

**Comment by thaJeztah** [comment]
@tonistiigi ptal

File: daemon/images/image_builder.go [hunk 3]
```diff
@@ -161,7 +167,28 @@ func (i *ImageService) pullForBuilder(ctx context.Context, name string, authConf
 	if err := i.pullImageWithReference(ctx, ref, platform, nil, pullRegistryAuth, output); err != nil {
 		return nil, err
 	}
-	return i.GetImage(name, platform)
+
+	img, err := i.GetImage(name, platform)
+	if errdefs.IsNotFound(err) && img != nil && platform != nil {
+		imgPlat, parseErr := platforms.Parse(path.Join(img.OS, img.BaseImgArch(), img.BaseImgVariant()))
```
Comment (tonistiigi) [inline daemon/images/image_builder.go:173]: Just create struct directly instead of parsing string

**Review by tonistiigi — APPROVED** [review]

**Comment by thaJeztah** [comment]
@cpuguy83 did you rebase? I see `TestBuildUserNamespaceValidateCapabilitiesAreV2` is failing, and that should be fixed by https://github.com/moby/moby/pull/41865

**Comment by cpuguy83** [comment]
Rebased.

**Review by thaJeztah — APPROVED** [review]
LGTM

s390x failure is unrelated

**Pull request merged**